File: packages/marimo-studio/src/marimo_studio/_projections/resolution.py
"""Authorize one concrete mount request against the saved notebook graph.

Resolution starts from a mount site recorded in the published artifact. It
checks the requested kind, allowed target, selector syntax, and browser-visible
size limits, then requires one unambiguous source cell or variable. The result
includes the upstream cells Marimo must execute before that target is ready.

Missing, ambiguous, disallowed, and oversized requests return stable error
codes before they reach a kernel. Runtime configuration derives target records
and limits from this module, while browser-evidence verification resolves each
observed instance through it.
"""

from __future__ import annotations

from dataclasses import dataclass

from marimo_studio._notebook.records import CellRef
from marimo_studio._projections.records import ValuePathStep, ValueReference
from marimo_studio._projections.symbol_graph import NotebookSymbolGraph
from marimo_studio._projections.values import (
    MAX_VALUE_PATH_STEPS,
    MAX_VALUE_REFERENCE_BYTES,
    UnpairedUTF16SurrogateError,
    normalize_utf16_surrogate_pairs,
    parse_value_reference,
)
from marimo_studio.view_providers import (
    MountDeclaration,
    ProjectionKind,
    SourceLocation,
)
from marimo_studio.view_providers._targets import validate_projection_target

MAX_PROJECTION_INSTANCE_ID_BYTES = 256
MAX_ACTIVE_PROJECTION_INSTANCES = 512
MAX_UNIQUE_CELL_TARGETS = 256
MAX_UNIQUE_OUTPUT_TARGETS = 100
MAX_UNIQUE_VALUE_TARGETS = 100
PROJECTION_UNPAIRED_SURROGATE_CODE = "projection-unpaired-surrogate"


class ProjectionResolutionError(ValueError):
    """One projection request cannot resolve inside its presentation."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


@dataclass(frozen=True)
class ProjectionRequest:
    """One mounted browser host requesting a notebook projection target."""

    site_id: str
    instance_id: str
    target: str

    def __post_init__(self) -> None:
        values = {"projection instance ID": self.instance_id}
        for label, value in values.items():
            if not isinstance(value, str) or not value:
                raise ProjectionResolutionError(
                    "invalid-projection-request",
                    f"The {label} must be a non-empty string.",
                )
        if not isinstance(self.site_id, str):
            raise ProjectionResolutionError(
                "invalid-projection-request",
                "The projection site ID must be a string.",
            )
        if not isinstance(self.target, str):
            raise ProjectionResolutionError(
                "invalid-projection-request",
                "The projection target must be a string.",
            )
        try:
            instance_id = normalize_utf16_surrogate_pairs(self.instance_id)
            target = normalize_utf16_surrogate_pairs(self.target)
        except UnpairedUTF16SurrogateError as error:
            raise ProjectionResolutionError(
                PROJECTION_UNPAIRED_SURROGATE_CODE,
                "Projection targets and instance IDs require well-formed Unicode.",
            ) from error
        object.__setattr__(self, "instance_id", instance_id)
        object.__setattr__(self, "target", target)
        if len(self.instance_id.encode("utf-8")) > MAX_PROJECTION_INSTANCE_ID_BYTES:
            raise ProjectionResolutionError(
                "projection-instance-id-too-large",
                "The projection instance ID exceeds the byte limit.",
            )
        if len(self.target.encode("utf-8")) > MAX_VALUE_REFERENCE_BYTES:
            raise ProjectionResolutionError(
                "projection-target-too-large",
                f"The projection target exceeds {MAX_VALUE_REFERENCE_BYTES} bytes.",
            )

    @classmethod
    def from_dict(
        cls,
        value: object,
    ) -> ProjectionRequest:
        if not isinstance(value, dict) or set(value) != {
            "siteId",
            "instanceId",
            "target",
        }:
            raise ProjectionResolutionError(
                "invalid-projection-request",
                "A projection request must contain siteId, instanceId, and target.",
            )
        site_id = value.get("siteId")
        instance_id = value.get("instanceId")
        target = value.get("target")
        if not all(isinstance(item, str) for item in (site_id, instance_id, target)):
            raise ProjectionResolutionError(
                "invalid-projection-request",
                "Projection request identifiers and targets must be strings.",
            )
        assert isinstance(site_id, str)
        assert isinstance(instance_id, str)
        assert isinstance(target, str)
        return cls(
            site_id=site_id,
            instance_id=instance_id,
            target=target,
        )


@dataclass(frozen=True)
class ResolvedProjection:
    """A projection request joined to its stable notebook producer."""

    request: ProjectionRequest
    kind: ProjectionKind
    source: SourceLocation
    producer: CellRef
    variable: str | None
    selector_path: tuple[ValuePathStep, ...]
    dependency_closure: tuple[CellRef, ...]

    def selector_spec(self) -> tuple[str, tuple[tuple[str, str | int], ...]]:
        if self.variable is None:
            raise ValueError("Cell projections do not have selector specs")
        return (
            self.variable,
            tuple((step.kind, step.value) for step in self.selector_path),
        )

    def to_dict(self) -> dict[str, object]:
        return {
            "siteId": self.request.site_id,
            "instanceId": self.request.instance_id,
            "kind": self.kind,
            "target": self.request.target,
            "source": self.source.to_dict(),
            "producer": str(self.producer),
            "variable": self.variable,
            "selectorPath": [
                {"kind": step.kind, "value": step.value} for step in self.selector_path
            ],
            "dependencyClosure": [str(ref) for ref in self.dependency_closure],
        }


def _site_for(
    sites: tuple[MountDeclaration, ...],
    request: ProjectionRequest,
) -> MountDeclaration:
    matches = tuple(site for site in sites if site.id == request.site_id)
    if len(matches) != 1:
        raise ProjectionResolutionError(
            "projection-site-not-found",
            f"Projection site {request.site_id!r} is unavailable.",
        )
    site = matches[0]
    return site


def _authorize_target(site: MountDeclaration, target: str) -> None:
    if site.allowed_targets is not None and target not in site.allowed_targets:
        raise ProjectionResolutionError(
            "projection-target-not-allowed",
            f"Projection target {target!r} is not allowed by mount {site.id!r}.",
        )


def validate_mount_declaration(site: MountDeclaration) -> None:
    """Validate one provider-owned site before it enters a presentation."""
    targets = site.allowed_targets
    if targets is None:
        return
    if not targets or len(set(targets)) != len(targets):
        raise ProjectionResolutionError(
            "projection-mount-invalid",
            f"Projection mount {site.id!r} has invalid allowed targets.",
        )
    for target in targets:
        if not target or len(target.encode("utf-8")) > MAX_VALUE_REFERENCE_BYTES:
            raise ProjectionResolutionError(
                "projection-mount-invalid",
                f"Projection mount {site.id!r} contains an invalid target.",
            )


def _unique_ref(
    candidates: tuple[CellRef, ...],
    *,
    target: str,
    missing_code: str,
    ambiguous_code: str,
    noun: str,
) -> CellRef:
    if not candidates:
        raise ProjectionResolutionError(
            missing_code,
            f"{noun} {target!r} does not resolve in the notebook.",
        )
    if len(candidates) != 1:
        raise ProjectionResolutionError(
            ambiguous_code,
            f"{noun} {target!r} has multiple notebook producers.",
        )
    return candidates[0]


def resolve_projection(
    graph: NotebookSymbolGraph,
    sites: tuple[MountDeclaration, ...],
    request: ProjectionRequest,
) -> ResolvedProjection:
    """Resolve one target through its artifact site and notebook graph."""
    site = _site_for(sites, request)
    validate_mount_declaration(site)
    if not request.target:
        raise ProjectionResolutionError(
            "projection-target-empty",
            "A mounted projection target must not be empty.",
        )
    try:
        validate_projection_target(site.kind, request.target)
    except UnpairedUTF16SurrogateError as error:
        raise ProjectionResolutionError(
            PROJECTION_UNPAIRED_SURROGATE_CODE,
            "Projection targets and instance IDs require well-formed Unicode.",
        ) from error
    except ValueError as error:
        raise ProjectionResolutionError(
            "projection-target-invalid",
            f"Projection target {request.target!r} is invalid: {error}",
        ) from error
    _authorize_target(site, request.target)
    reference: ValueReference | None = None
    if site.kind == "cell":
        producer = _unique_ref(
            graph.cell_targets.get(request.target, ()),
            target=request.target,
            missing_code="projection-cell-not-found",
            ambiguous_code="projection-cell-ambiguous",
            noun="Cell target",
        )
    else:
        try:
            reference = parse_value_reference(request.target)
        except UnpairedUTF16SurrogateError as error:
            raise ProjectionResolutionError(
                PROJECTION_UNPAIRED_SURROGATE_CODE,
                "Projection targets and instance IDs require well-formed Unicode.",
            ) from error
        except ValueError as error:
            raise ProjectionResolutionError(
                "projection-target-invalid",
                f"Projection target {request.target!r} is invalid: {error}",
            ) from error
        variable = graph.variables.get(reference.variable)
        producer = _unique_ref(
            variable.producers if variable is not None else (),
            target=reference.variable,
            missing_code=f"projection-{site.kind}-variable-not-found",
            ambiguous_code=f"projection-{site.kind}-variable-ambiguous",
            noun="Notebook variable",
        )
    return ResolvedProjection(
        request=request,
        kind=site.kind,
        source=site.source,
        producer=producer,
        variable=reference.variable if reference is not None else None,
        selector_path=reference.path if reference is not None else (),
        dependency_closure=graph.dependency_closure(producer),
    )


def _target_record(
    graph: NotebookSymbolGraph,
    candidates: tuple[CellRef, ...],
) -> dict[str, object] | None:
    if not candidates:
        return None
    if len(candidates) > 1:
        return {"status": "ambiguous"}
    producer = next(iter(candidates))
    return {
        "status": "ready",
        "producer": str(producer),
        "dependencyClosure": [
            str(reference) for reference in graph.dependency_closure(producer)
        ],
    }


def projection_targets(
    graph: NotebookSymbolGraph,
    sites: tuple[MountDeclaration, ...],
) -> dict[str, object]:
    """Return the target records required by one artifact's mount declarations."""
    cell_targets: set[str] = set()
    variable_targets: set[str] = set()
    all_cells = False
    all_variables = False
    for site in sites:
        if site.allowed_targets is None:
            if site.kind == "cell":
                all_cells = True
            else:
                all_variables = True
            continue
        if site.kind == "cell":
            cell_targets.update(site.allowed_targets)
            continue
        for target in site.allowed_targets:
            variable_targets.add(parse_value_reference(target).variable)
    if all_cells:
        cell_targets.update(graph.cell_targets)
    if all_variables:
        variable_targets.update(graph.variables)

    cells = {
        target: record
        for target in sorted(cell_targets)
        if (record := _target_record(graph, graph.cell_targets.get(target, ())))
        is not None
    }
    variables = {
        target: record
        for target in sorted(variable_targets)
        if (
            record := _target_record(
                graph,
                (
                    graph.variables[target].producers
                    if target in graph.variables
                    else ()
                ),
            )
        )
        is not None
    }
    return {"cells": cells, "variables": variables}


def projection_policy() -> dict[str, int]:
    """Return browser-visible limits enforced by the Python resolver."""
    return {
        "maxActiveInstances": MAX_ACTIVE_PROJECTION_INSTANCES,
        "maxUniqueCellTargets": MAX_UNIQUE_CELL_TARGETS,
        "maxUniqueOutputTargets": MAX_UNIQUE_OUTPUT_TARGETS,
        "maxUniqueValueTargets": MAX_UNIQUE_VALUE_TARGETS,
        "maxTargetBytes": MAX_VALUE_REFERENCE_BYTES,
        "maxPathSteps": MAX_VALUE_PATH_STEPS,
        "maxInstanceIdBytes": MAX_PROJECTION_INSTANCE_ID_BYTES,
    }

File: packages/presentation/src/projections/resolution.ts
import type {
  MountDeclaration,
  ProjectionKind,
  ProjectionRequest,
  ProjectionTarget,
  SourceLocation,
} from "@marimo-studio/protocol/projections";
import type { RuntimeConfig } from "@marimo-studio/protocol/runtime-config";

import {
  PROJECTION_UNPAIRED_SURROGATE_CODE,
  projectionStringIsWellFormed,
} from "@marimo-studio/protocol/projections";
import { z } from "zod";

import { ownRecordValue } from "../records.ts";
import { isArtifactProjectionHost } from "./artifact-host.ts";
import { projectionKindForHost, projectionTargetForHost } from "./identity.ts";
import { notifyProjectionResolutionStale } from "./staleness.ts";

export interface SelectorPathStep {
  readonly kind: "attribute" | "item";
  readonly value: string | number;
}

export interface RuntimeProjectionRequest extends ProjectionRequest {
  readonly kind: ProjectionKind;
}

export interface ResolvedProjection {
  readonly request: RuntimeProjectionRequest;
  readonly site: MountDeclaration;
  readonly producer: string;
  readonly variable: string | null;
  readonly selectorPath: readonly SelectorPathStep[];
  readonly dependencyClosure: readonly string[];
  readonly runtimeCellId: string | undefined;
  readonly bindingsStale: boolean;
}

export interface ProjectionResolutionFailure {
  readonly code: string;
  readonly message: string;
  readonly siteId: string;
  readonly instanceId: string;
  readonly kind: ProjectionKind;
  readonly target: string;
  readonly source?: SourceLocation;
}

export type ProjectionResolution =
  | { readonly ok: true; readonly value: ResolvedProjection }
  | { readonly ok: false; readonly error: ProjectionResolutionFailure };

type ProjectionFailure = Extract<ProjectionResolution, { readonly ok: false }>;

interface ParsedSelector {
  readonly variable: string;
  readonly path: readonly SelectorPathStep[];
}

type ParsedSelectorResult =
  | { readonly ok: true; readonly value: ParsedSelector }
  | { readonly ok: false; readonly code?: string; readonly message: string };

type ProducerResolution =
  | { readonly ok: true; readonly value: Extract<ProjectionTarget, { status: "ready" }> }
  | ProjectionFailure;

const IDENTIFIER = /^[A-Za-z_][A-Za-z0-9_]*/;
const INDEX = /^(?:0|[1-9][0-9]*)/;
const HOST_SELECTOR = "marimo-cell, marimo-output, [mo-value]";

const failure = (
  request: RuntimeProjectionRequest,
  code: string,
  message: string,
  site?: MountDeclaration,
): ProjectionFailure => ({
  ok: false,
  error: {
    code,
    message,
    siteId: request.siteId,
    instanceId: request.instanceId,
    kind: request.kind,
    target: request.target,
    source: site?.source,
  },
});

const parseSelector = (target: string, maxPathSteps: number): ParsedSelectorResult => {
  const source = target.trim();
  const root = IDENTIFIER.exec(source);
  if (root === null) {
    return {
      ok: false,
      message: "A value target must start with a notebook variable name.",
    };
  }
  const path: SelectorPathStep[] = [];
  let position = root[0].length;
  while (position < source.length) {
    const suffix = source.slice(position);
    if (suffix.startsWith(".")) {
      const selected = IDENTIFIER.exec(suffix.slice(1));
      if (selected === null) {
        return { ok: false, message: "Dot selection requires an object key." };
      }
      if (selected[0].startsWith("_")) {
        return { ok: false, message: "Private attribute selection is unavailable." };
      }
      path.push({ kind: "attribute", value: selected[0] });
      position += selected[0].length + 1;
    } else if (suffix.startsWith("[")) {
      const content = suffix.slice(1);
      const selectedIndex = INDEX.exec(content);
      if (selectedIndex !== null) {
        const index = Number(selectedIndex[0]);
        if (!Number.isSafeInteger(index)) {
          return {
            ok: false,
            message: "Bracket indexes must be JavaScript safe integers.",
          };
        }
        path.push({ kind: "item", value: index });
        position += selectedIndex[0].length + 1;
      } else if (content.startsWith('"')) {
        let selected: unknown;
        let consumed = 0;
        for (let end = 2; end <= content.length; end += 1) {
          try {
            selected = JSON.parse(content.slice(0, end));
            consumed = end;
            break;
          } catch {
            // Continue until one complete JSON string has been consumed.
          }
        }
        const selectedKey = z.string().safeParse(selected);
        if (!selectedKey.success) {
          return {
            ok: false,
            message: "Bracket object keys must be JSON strings.",
          };
        }
        if (!projectionStringIsWellFormed(selectedKey.data)) {
          return {
            ok: false,
            code: PROJECTION_UNPAIRED_SURROGATE_CODE,
            message: "Projection targets and instance IDs require well-formed Unicode.",
          };
        }
        path.push({ kind: "item", value: selectedKey.data });
        position += consumed + 1;
      } else {
        return {
          ok: false,
          message: "Brackets require a non-negative integer or JSON string.",
        };
      }
      if (source[position] !== "]") {
        return {
          ok: false,
          message: "Bracket selection requires a closing ].",
        };
      }
      position += 1;
    } else {
      return {
        ok: false,
        message: "A value target may contain dot selection and bracket indexing.",
      };
    }
    if (path.length > maxPathSteps) {
      return {
        ok: false,
        message: `A value target may contain at most ${maxPathSteps} path steps.`,
      };
    }
  }
  return { ok: true, value: { variable: root[0], path } };
};

const oneProducer = (
  request: RuntimeProjectionRequest,
  target: ProjectionTarget | undefined,
  noun: string,
  site: MountDeclaration,
): ProducerResolution => {
  const codeNoun = noun === "cell" ? "" : `-${noun}`;
  if (target === undefined) {
    return failure(
      request,
      `projection-${request.kind}${codeNoun}-not-found`,
      `${noun === "cell" ? "Cell target" : "Notebook variable"} ${JSON.stringify(
        request.target,
      )} does not resolve in this notebook.`,
      site,
    );
  }
  if (target.status === "ambiguous") {
    return failure(
      request,
      `projection-${request.kind}${codeNoun}-ambiguous`,
      `${noun === "cell" ? "Cell target" : "Notebook variable"} ${JSON.stringify(
        request.target,
      )} has multiple notebook producers.`,
      site,
    );
  }
  return { ok: true, value: target };
};

export const resolveProjection = (
  config: RuntimeConfig,
  request: RuntimeProjectionRequest,
  context?: ProjectionResolutionContext,
): ProjectionResolution => {
  if (
    !projectionStringIsWellFormed(request.instanceId) ||
    !projectionStringIsWellFormed(request.target)
  ) {
    return failure(
      request,
      PROJECTION_UNPAIRED_SURROGATE_CODE,
      "Projection targets and instance IDs require well-formed Unicode.",
    );
  }
  if (
    new TextEncoder().encode(request.instanceId).length > config.projectionPolicy.maxInstanceIdBytes
  ) {
    return failure(
      request,
      "projection-instance-id-too-large",
      "The projection instance identifier exceeds the runtime limit.",
    );
  }
  if (new TextEncoder().encode(request.target).length > config.projectionPolicy.maxTargetBytes) {
    return failure(
      request,
      "projection-target-too-large",
      "The projection target exceeds the runtime limit.",
    );
  }
  const matches =
    context?.site(request.siteId) ?? config.mounts.filter((site) => site.id === request.siteId);
  if (matches.length !== 1) {
    return failure(
      request,
      "projection-site-not-found",
      request.siteId
        ? `Projection site ${JSON.stringify(request.siteId)} is unavailable.`
        : "The mounted projection has no provider source-site identity.",
    );
  }
  const site = matches[0];
  if (site.kind !== request.kind) {
    return failure(
      request,
      "projection-kind-mismatch",
      `Projection site ${JSON.stringify(site.id)} cannot request ${request.kind} targets.`,
      site,
    );
  }
  if (request.target.length === 0) {
    return failure(
      request,
      "projection-target-empty",
      "A mounted projection target must not be empty.",
      site,
    );
  }
  if (site.allowedTargets !== null && !site.allowedTargets.includes(request.target)) {
    return failure(
      request,
      "projection-target-not-allowed",
      `Projection target ${JSON.stringify(request.target)} is not allowed by this mount.`,
      site,
    );
  }

  let producer: ProducerResolution;
  let variable: string | null = null;
  let selectorPath: readonly SelectorPathStep[] = [];
  if (request.kind === "cell") {
    producer = oneProducer(
      request,
      ownRecordValue(config.projectionTargets.cells, request.target),
      "cell",
      site,
    );
  } else {
    const parsed = parseSelector(request.target, config.projectionPolicy.maxPathSteps);
    if (!parsed.ok) {
      return failure(request, parsed.code ?? "projection-target-invalid", parsed.message, site);
    }
    variable = parsed.value.variable;
    selectorPath = parsed.value.path;
    producer = oneProducer(
      request,
      ownRecordValue(config.projectionTargets.variables, parsed.value.variable),
      "variable",
      site,
    );
  }
  if (!producer.ok) {
    return producer;
  }
  const producerRef = producer.value.producer;
  const dependencyClosure = producer.value.dependencyClosure;
  const runtimeBindings = config.runtimeBindings.cellRefs;
  const bindingsStale =
    Object.keys(runtimeBindings).length > 0 &&
    dependencyClosure.some((reference) => ownRecordValue(runtimeBindings, reference) === undefined);
  return {
    ok: true,
    value: {
      request,
      site,
      producer: producerRef,
      variable,
      selectorPath,
      dependencyClosure,
      runtimeCellId: ownRecordValue(runtimeBindings, producerRef),
      bindingsStale,
    },
  };
};

export interface ProjectionResolutionContext {
  activeIndex(host: Element): number | undefined;
  hosts(): readonly Element[];
  site(id: string): readonly MountDeclaration[];
  targetRank(kind: ProjectionKind, target: string): number | undefined;
}

export const createProjectionResolutionContext = (
  config: RuntimeConfig,
  root: ParentNode,
): ProjectionResolutionContext => {
  const sites = new Map<string, MountDeclaration[]>();
  config.mounts.forEach((site) => {
    const matches = sites.get(site.id) ?? [];
    matches.push(site);
    sites.set(site.id, matches);
  });
  const hosts = Array.from(root.querySelectorAll(HOST_SELECTOR)).filter(isArtifactProjectionHost);
  const activeIndexes = new Map(hosts.map((host, index) => [host, index]));
  const ranks: Record<ProjectionKind, Map<string, number>> = {
    cell: new Map(),
    output: new Map(),
    value: new Map(),
  };
  hosts.forEach((host, index) => {
    const kind = projectionKindForHost(host);
    const target = projectionTargetForHost(host, kind);
    const resolved = resolveProjection(config, {
      siteId: host.getAttribute("data-marimo-studio-site")?.trim() ?? "",
      instanceId: `quota-${index}`,
      kind,
      target,
    });
    if (!resolved.ok) {
      return;
    }
    const selected = ranks[kind];
    if (!selected.has(target)) {
      selected.set(target, selected.size);
    }
  });
  return {
    activeIndex: (host) => activeIndexes.get(host),
    hosts: () => hosts,
    site: (id) => sites.get(id) ?? [],
    targetRank: (kind, target) => ranks[kind].get(target),
  };
};

export const resolveHostProjection = (
  config: RuntimeConfig,
  host: Element,
  request: RuntimeProjectionRequest,
  context: ProjectionResolutionContext = createProjectionResolutionContext(
    config,
    host.ownerDocument,
  ),
): ProjectionResolution => {
  const resolved = resolveProjection(config, request, context);
  if (!resolved.ok) {
    return resolved;
  }
  if (resolved.value.bindingsStale) {
    notifyProjectionResolutionStale(config.projectionRevision);
  }
  const activeIndex = context.activeIndex(host);
  if (activeIndex === undefined || activeIndex >= config.projectionPolicy.maxActiveInstances) {
    return failure(
      request,
      "projection-instance-limit",
      `A presentation may mount at most ${config.projectionPolicy.maxActiveInstances} projection instances.`,
      resolved.value.site,
    );
  }
  const limits: Record<ProjectionKind, number> = {
    cell: config.projectionPolicy.maxUniqueCellTargets,
    output: config.projectionPolicy.maxUniqueOutputTargets,
    value: config.projectionPolicy.maxUniqueValueTargets,
  };
  const limit = ownRecordValue(limits, request.kind);
  if (limit === undefined) {
    return failure(
      request,
      "projection-kind-mismatch",
      "The mounted projection kind is unavailable.",
      resolved.value.site,
    );
  }
  const targetRank = context.targetRank(request.kind, request.target);
  if (targetRank !== undefined && targetRank >= limit) {
    return failure(
      request,
      `projection-${request.kind}-target-limit`,
      `A presentation may mount at most ${limit} unique ${request.kind} targets.`,
      resolved.value.site,
    );
  }
  return resolved;
};
